audio/out/ao_oss.c

```diff
@@ -0,0 +1,439 @@
+    ao_control_vol_t *vol = (ao_control_vol_t *)arg;
+    int v;
+
+    if (p->dsp_fd < 0)
+        return CONTROL_ERROR;
+
+    switch (cmd) {
+    case AOCONTROL_GET_VOLUME:
+        if (ioctl(p->dsp_fd, SNDCTL_DSP_GETPLAYVOL, &v) == -1)
+            return CONTROL_ERROR;
+        vol->right = ((v & 0xff00) >> 8);
+        vol->left = (v & 0x00ff);
+        return CONTROL_OK;
+    case AOCONTROL_SET_VOLUME:
+        v = ((int)vol->right << 8) | (int)vol->left;
+        if (ioctl(p->dsp_fd, SNDCTL_DSP_SETPLAYVOL, &v) == -1)
+            return CONTROL_ERROR;
+        return CONTROL_OK;
+    }
+
+    return CONTROL_UNKNOWN;
+}
+
+static void reset(struct ao *ao)
+{
+    struct priv *p = ao->priv;
+
+    p->playing = false;
+    if (ioctl(p->dsp_fd, SNDCTL_DSP_RESET, NULL) == 0) {
+        set_pause(ao, true);
+    } else {
+        MP_WARN_IOCTL_ERR(ao);
+        if (p->resetting)
+            return; /* Avoid infinite loop: init() -> reset() -> init()... */
+        p->resetting = true;
+        MP_WARN(ao, "Force reinitialize audio device.\n");
+        uninit(ao);
+        init(ao);
+        p->resetting = false;
+    }
+}
+
+static bool set_pause(struct ao *ao, bool pause)
+{
+    struct priv *p = ao->priv;
+    int trig = 0; /* Both directions disabled */
+
+    if (!pause) {
+        trig |= (PCM_ENABLE_INPUT | PCM_ENABLE_OUTPUT); /* Both directions enabled */
+    }
+    if (ioctl(p->dsp_fd, SNDCTL_DSP_SETTRIGGER, &trig) != 0) {
+        p->playing = false;
+        MP_WARN_IOCTL_ERR(ao);
+    }
+
+    return true;
+}
+
+static void start(struct ao *ao)
+{
+    struct priv *p = ao->priv;
+
+    p->playing = true;
+    set_pause(ao, false);
+}
+
+static bool audio_write(struct ao *ao, void **data, int samples)
+{
+    struct priv *p = ao->priv;
+    const size_t size = (samples * ao->sstride);
+    ssize_t rc;
+
+    if (size == 0)
+        return true;
+
+    rc = write(p->dsp_fd, data[0], size);
+    if (rc == -1) {
+        MP_WARN(ao, "audio_write: write() fail, err = %i: %s\n",
+            errno, strerror(errno));
+        return false;
+    }
+    if ((size_t)rc != size) {
+        p->playing = false;
+        MP_WARN(ao, "audio_write: unexpected partial write: required: %zu, written: %zu\n",
+            size, (size_t)rc);
+        return false;
+    }
+
+    return true;
+}
+
+static void get_state(struct ao *ao, struct mp_pcm_state *state)
+{
+    struct priv *p = ao->priv;
+    audio_buf_info info;
+    int odelay;
+
+    /* free_samples, queued_samples */
+    if (ioctl(p->dsp_fd, SNDCTL_DSP_GETOSPACE, &info) != -1) {
+        state->free_samples = (info.bytes / ao->sstride);
+        state->queued_samples = (ao->device_buffer - state->free_samples);
+    } else {
+        state->free_samples = state->queued_samples = 0;
+        p->playing = false;
+        MP_WARN_IOCTL_ERR(ao);
+    }
+    /* delay */
+    if (ioctl(p->dsp_fd, SNDCTL_DSP_GETODELAY, &odelay) != -1) {
+        state->delay = (odelay / (double)ao->bps);
```

Comment:
also needs to be divided by number of channels, as per http://manuals.opensound.com/developer/SNDCTL_DSP_GETODELAY.html
>  The delay is returned in bytes and it can be converted to seconds by dividing it by sample_rate*num_of_channels*sample_size_in_bytes.

Though whether `ao->bps` is per-channel or not is something that apparently not even mpv itself agrees on; ao.c sets it to `ao->bps = ao->samplerate * ao->sstride;` whereas ao_pcm.c does `ao->bps = ao->channels.num * ao->samplerate * af_fmt_to_bytes(ao->format);`

Reply:
Good catch!